samples/subsys/nvs/src/main.c

```diff
@@ -0,0 +1,154 @@
+/*
+ * NVS Sample for zephyr using high level API, this sample stores a string
+ * (address), a binary blob (key) and a long integer (reboot counter) in flash
+ * the samples reboots every 5s to show the data that was stored in flash. It
+ * stops after 82 reboots (when reboot counter is 80 the write of reboot counter
+ * triggers a sector change which copies the data of address, key and reboot
+ * counter before erasing the flash). As a result of the rotation the history
+ * data of reboot counter is reduced in size but the address and key data is
+ * kept.
+ *
+ * Copyright (c) 2018 Laczen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <zephyr.h>
+#include <misc/reboot.h>
+#include <board.h>
+#include <device.h>
+#include <string.h>
+#include <nvs/nvs.h>
+
+#define STORAGE_OFFSET 0x3F000
+#define STORAGE_MAGIC 0x4d455348 /* hex for "MESH" */
+#define SECTOR_SIZE 0x400
+#define SECTOR_COUNT 2
+
+static struct nvs_fs fs = {
+	.sector_size = SECTOR_SIZE,
+	.sector_count = SECTOR_COUNT,
+	.offset = STORAGE_OFFSET,
+};
+
+/* 1000 msec = 1 sec */
+#define SLEEP_TIME      1000
+/* maximum reboot counts, make high enough to trigger sector change (buffer */
+/* rotation). */
+#define MAX_REBOOT 82
+
+void main(void)
+{
+	int rc = 0, cnt = 0;
+	struct nvs_entry entry;
+	char buf[] = "192.168.1.1";
+	u8_t key[8] = { 0xFF, 0xFE, 0xFD, 0xFC, 0xFB, 0xFA, 0xF9, 0xF8 };
+	u32_t reboot_counter = 0;
+
+	rc = nvs_init(&fs, FLASH_DEV_NAME, STORAGE_MAGIC);
+	if (rc) {
+		printk("Flash Init failed\n");
+	}
+
+	/* entry with id=1 is used to store an address, lets see if we can
+	 * read it from flash
+	 */
+	entry.id = 1;
+	rc = nvs_read(&fs, &entry, &buf);
```

Comment:
Won't this overflow `buf` if anything larger than `sizeof(buf)` is read from the flash? Could be by malice, could be by a hardware issue, could even be a different IP address.

Reply:
@lpereira You are right, as a read operation is done first a buffer overflow might occur if there is data (with id=1) in flash that is larger than buf. I should have used another variable.

Reply:
@lpereira I have been looking into this and I have come up with an alternative solution: what if I would change nvs_read(&fs, &entry, &buf) to nvs_read(&fs, &entry, &but, int maxreadlength). Then by passing the size of the buffer to nvs_read it is not possible to overflow the buffer. The data that would be read is of size min(what is in flash, maxreadlength). It is then also possible to compare the len of what is in flash with the buffer size. If the buffer size is to small, adapt and call nvs_read again. What is your opinion ?

Reply:
Yes, that would work (although I would prefer a size_t instead of int) as far as preventing buffer overflows.